### ISSUE TITLE
Class Specific Attention in Keras

### ISSUE THREAD

**Issue by siddheshk** [fchollet/keras#5330]
I am trying to implement a model that has class specific attention (different attention for each class label). So if there are k classes, there will be k such attention layers. Also, only the attention layer is different. All other layers are common.

I have been able to do this by defining k different models, each having a separate attention layer. But the problem with this is that I am restricted to training this by passing one training instance at a time (As I need to select the model depending on the value of the training label).

Is it possible to do so in a batch fashion? Like having one single model that selects the correct attention layer and updates the weights of only that corresponding layer?

I tried out something like this
```python
class AttentionLayer(Layer):
    def build(self, input_shape):
        input_dim = input_shape[0][-1]
        self.W_w = self.add_weight(shape=(self.num_classes, input_dim, input_dim), initializer='glorot_uniform', trainable=True, name='{}_W_w'.format(self.name))
        self.b_w = self.add_weight(shape=(self.num_classes, input_dim,), initializer='zero', trainable=True, name='{}_b_w'.format(self.name))
        self.u_w = self.add_weight(shape=(self.num_classes, input_dim,), initializer='glorot_uniform', trainable=True, name='{}_u_w'.format(self.name))
        super(AttentionLayer, self).build(input_shape)

    def call(self, inputs, mask=None):
        x = inputs[0]
        label = inputs[1][0][0]
        u_it = K.tanh(K.dot(x, self.W_w[label]) + self.b_w[label])
        return u_it
```
But in this case, when I pass only a single training instance through the model, the entire `self.W_w`  matrix is updated (Even the slices which weren't used in computation). 

Is there a way to achieve this?

Please make sure that the boxes below are checked before you submit your issue. If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [ ] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by unrealwill** [comment]
Hello,
I guess you can use three EmbeddingLayer which would select your attention weights which you can then combine appropriately.

Or you can for just a little more than the price you pay currently, have the label be continous (as a softmax) rather than discrete, so gradient could flow through them.

To not update the full matrix you are dependent on the autodiff framework, and it's implementation of the adjoint operation. I don't know if they did such optimizations for the embedding layer but I suppose so because it's the specific gather op which is used.

In a more generic fashion using sparse matrix, could lead to not having to update the whole matrix provided the matrix is sparse, but you are yet again dependent of a lower level. You can have a look at how to implement custom op in either theano or tensorflow on your own if needed.

**Comment by siddheshk** [comment]
Thanks for your help @unrealwill .
Will using embedding layers train similarly?
Also I didn't quite understand how continuous labels can be used. Could you provide me with an example?
Thanks.

**Comment by unrealwill** [comment]
For the use of embedding layers : 
inp = Input() #your layer which you will extract from
labelinp = Input() #some discrete input
embW = Reshape( (input_dim,input_dim )) ( Embedding( num_class, input_dim*inp_dim)(labelinp))
embB = Embedding( num_class, input_dim) (labelinp)
embU = Embedding( num_class,input_dim) (labelinp) #I don't know how you use it currently in your code it's ununsed
customMerge = Merge( merge = lambda x: K.dot( x[0],x[1] )+ x[2]  merge_output_shape=... )([inp,embW,embB,embU])

It could benefit from optimizations of the embedding layers and (maybe) not update unnecessary weights.

What I meant about the continuous label was a disctinct point (a possible extension of your question). 
For almost the same computational cost you currently do, you can have a "soft" attention class mechanism
where instead of W[label] you have sum( classweight * W[:,:,classlabel] )
which allows you to have a continous either one_hot encoded (in which case it's less performant than what you were doing), or which may come from a softmax layer, and may allow to handle uncertainty upon the label.
labelInp = Input( batch_shape=(None,nb_class) )   instead of labelInp = Input( batch_shape=(None,1) )
